lms/djangoapps/courseware/views.py

```diff
@@ -1230,3 +1236,104 @@ def course_survey(request, course_id):
         redirect_url=redirect_url,
         is_required=course.course_survey_required,
     )
+
+
+@require_POST
+def generate_user_cert(request, course_id):
+    """
+    It will add the add-cert request into the xqueue.
+
+     Arguments:
+        request (django request object):  the HTTP request object that triggered this view function
+        course_id (unicode):  id associated with the course
+
+    Returns:
+        returns json response
+    """
+
+    if not request.user.is_authenticated():
+        log.info(u"Anon user trying to generate certificate for %s", course_id)
+        return JsonResponseBadRequest(_('You must be logged-in to generate certificate'))
+
+    student = request.user
+
+    # checking course id
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+
+    course = modulestore().get_course(course_key, depth=2)
+    if not course:
+        return JsonResponseBadRequest(_("Course is not valid"))
+
+    grade = grades.grade(student, request, course)
+
+    if not is_course_passed(course, grade):
+        return JsonResponseBadRequest(_("You failed to pass the course."))
+
+    xqueue = XQueueCertInterface()
+
+    certificate_status = certificate_downloadable_status(student, course_key)
+    if not certificate_status["is_downloadable"] and not certificate_status["is_generating"]:
+        ret = xqueue.add_cert(student, course_key, course=course)
+        log.info(
+            (
+                u"Added a certificate generation task to the XQueue "
+                u"for student %s in course '%s'. "
+                u"The new certificate status is '%s'."
+            ),
+            student.id,
+            unicode(course_key),
+            ret
+        )
+        return JsonResponse(_("Certificate generated."))
+    else:
+        log.warning(
+            (
+                u"user %s "
+                u"with course '%s'; "
+                u"not eligible.Reason is %s"
+            ),
+            student.id,
+            unicode(course_key), '',
+        )
+        # for any other status return bad request response
+        return JsonResponseBadRequest('')
+
+
+def is_course_passed(course, grade_summary):
```

Comment:
This function is used only in this module, so it should be called `_is_course_passed()` to mark it as private.

Also, I'd prefer if `grade_summary` were an optional argument set equal to `None` by default.  If not provided, this function can call `grades.grade()` directly instead of relying on the caller to do so.